### PR TITLE
Versions/0.0.2

### LINKED ISSUE #2: read about interrupts

### LINKED ISSUE #3: build basic GDT

### LINKED ISSUE #4: build basic IDT

### LINKED ISSUE #5: input output of ports

### LINKED ISSUE #6: build basic IRQ

### LINKED ISSUE #7: write a test interrupt

### LINKED ISSUE #8: write interrupts that are relevant to us

### LINKED ISSUE #9: test all interrupts

### PR DESCRIPTION
**Pull request by swimmy4days**
Finished basic interrupts, IDT, GDT, and very basic input

### CODE REVIEW

File: src/CLib/math/sqrt.c [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+The function calculates square root of a number and returns it
+float n - number to calculate its square root
+*/
+float sqrt(float n) 
+{ 
+    if(n < 0)
+        return -1;
+    float x = n; 
+    float y = 1; 
+    float e = 0.000001; /* e decides the accuracy level*/
```
Comment (ExhoAR22) [inline src/CLib/math/sqrt.c]: Your coding conventions should be consistent. In other files, say `pow.c`, you used `//` for single-line comments, yet here you use `/**/`.
I personally prefer `//` for single-line comments but you can choose whichever, as long as it's consistent throughout the codebase.

File: src/Kernel/interrupts/interrupts.s [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+.global irq0
+.global irq1
+
+.global irq0_handler
+.global irq1_handler
+
+.extern irq0_handler
+.extern irq1_handler
+
+irq0:
```
Comment (ExhoAR22) [inline src/Kernel/interrupts/interrupts.s]: Didn't we agree you'd implement these in C and use `__attribute__((interrupt))`?

File: Makefile [hunk 3]
```diff
@@ -2,44 +2,65 @@ include GCC_PATH
...
 
-buildObj = @echo "\033[36m[Compiling]\033[0m $(1).o" && $(CC) -c src/$(1)/$(1).c -o build/$(1).o $(CFLAGS)
+buildObj 	= @echo "\033[36m[Compiling]\033[0m $(if $(strip $(2)),$(2),$(1)).o" && $(CC) -c src/Kernel/$(1)/$(1).c -o build/$(if $(strip $(2)),$(2),$(1)).o $(CFLAGS)
+buildLibObj	= @echo "\033[36m[Compiling]\033[0m $(if $(strip $(2)),$(2),$(1)).o" && $(CC) -c src/CLib/$(1)/$(1).c -o build/$(if $(strip $(2)),$(2),$(1)).o $(CFLAGS)
+asmObj 		= @echo "\033[34m[Assembling]\033[0m $(1).o" && $(AS) src/kernel/$(1)/$(1).s -o build/$(1).o
 
 CC			= $(CUSTOM_GCC)/bin/i686-elf-gcc
 AS			= $(CUSTOM_GCC)/bin/i686-elf-as
-CFLAGS		= -ffreestanding -O2 -Wall -Wextra -std=gnu99
+CFLAGS		= -ffreestanding -O2 -Wall -Wextra -std=gnu99 -Isrc/include
```
Comment (ExhoAR22) [inline Makefile]: Why limit yourselves to an older C standard?
How about `-std=c18`?

File: src/CLib/stdio/printf.c [hunk 4]
```diff
@@ -1,26 +1,36 @@
...
+const char* data - string to print
+size_t length - length of string
+*/
 static bool print(const char *data, size_t length)
 {
-    const unsigned char *bytes = (const unsigned char *)data;
-    for (size_t i = 0; i < length; i++)
+    const unsigned char *bytes = (const unsigned char *)data; //assign string into bytes array of Ascii values
+    for (size_t i = 0; i < length; i++)                       //print each byte with putchar function
         if (putchar(bytes[i]) == EOF)
```
Comment (ExhoAR22) [inline src/CLib/stdio/printf.c]: Why would a function that prints out characters return `EOF`? The terminal should scroll once reached the end of the screen.

File: src/Kernel/interrupts/interrupts.c [hunk 5]
```diff
@@ -0,0 +1,5 @@
+//TODO - Find a better name for all the file in this directory
+//Gathers all the necessary files for the MakeFile
+
+#include "irq.c"
```
Comment (ExhoAR22) [inline src/Kernel/interrupts/interrupts.c:4]: Never include `.c` files, this is makes the code ugly and might cause problems in the future.

File: src/Kernel/io/io.s [hunk 6]
```diff
@@ -0,0 +1,16 @@
+/*Wrapper for the inb and outb functions*/
+.intel_syntax noprefix
+
+.global inb
+.global outb
+
+inb:
```
Comment (ExhoAR22) [inline src/Kernel/io/io.s]: These function could easily be implemented using inline assembly, like so:
```c
static inline void outb(uint16_t port, uint8_t val) {
    asm volatile("outb %0, %1" : : "a"(val), "Nd"(port));
}

static inline uint8_t inb(uint16_t port) {
    uint8_t ret;
    asm volatile("inb %1, %0" : "=a"(ret) : "Nd"(port));
    return ret;
}
```

File: src/include/idt.h [hunk 7]
```diff
@@ -0,0 +1,12 @@
+#pragma once
+
+#include "io.h"
+#ifdef __cplusplus
```
Comment (ExhoAR22) [inline src/include/idt.h]: Are these necessary? Do you ever intend to use C++ in this project?

File: src/kernel/boot.s [hunk 8]
```diff
@@ -91,17 +94,17 @@ _start:
...
 	gdt_data:
 		.word 0xFFFF
 		.word 0
 		
 		.byte 0
-		.byte 10010010
-		.byte 11001111
+		.byte 0x92 /*10010010b*/
+		.byte 0xCF /*11001111b*/
 		.byte 0
```
Comment (ExhoAR22) [inline src/kernel/boot.s]: Any reason the GDT is defined like so in assembly? Wouldn't it be better to make a `struct` for it?

File: src/CLib/stdio/putchar.c [hunk 9]
```diff
@@ -1,9 +1,11 @@
 #include "stdio.h"
 
+//Warper function, calls terminal_putchat
 int putchar(int ic)
```
Comment (ExhoAR22) [inline src/CLib/stdio/putchar.c]: Any reason C is an `int`?

**Review by ExhoAR22 — CHANGES_REQUESTED** [review]
Great job overall, fix the issues iv'e pointed out and your'e good to go!

File: src/Kernel/interrupts/interrupts.s (hunk 2 above)
Comment (yardendolev) [reply src/Kernel/interrupts/interrupts.s]: We had some issues using attribute. We get this error: sorry, unimplemented: 80387 instructions aren't allowed in an interrupt service routine

File: src/CLib/stdio/printf.c (hunk 4 above)
Comment (yardendolev) [reply src/CLib/stdio/printf.c]: According to the documentation putchar function returns EOF if an error is raised.

File: src/Kernel/interrupts/interrupts.c (hunk 5 above)
Comment (yardendolev) [reply src/Kernel/interrupts/interrupts.c:4]: Its for the makefile. Do you have a better solution for it?

File: src/kernel/boot.s (hunk 8 above)
Comment (swimmy4days) [reply src/kernel/boot.s]: Does it matter?
If it does, could you send us a link, we got a little lost.

File: src/Kernel/interrupts/interrupts.s (hunk 2 above)
Comment (ExhoAR22) [reply src/Kernel/interrupts/interrupts.s]: This is likely due to you using your own custom build of `gcc`, how about using mainline `clang`?

File: src/kernel/boot.s (hunk 8 above)
Comment (ExhoAR22) [reply src/kernel/boot.s]: Well, for starters, when using a struct other people can easily tell which field is which, while here we have a bunch of numbers without any names.
Also, it looks way better :sweat_smile:

File: src/CLib/stdio/printf.c (hunk 4 above)
Comment (ExhoAR22) [reply src/CLib/stdio/printf.c]: Mimicking the standard C library isn't such a great idea in my opinion - most other kernels don't do this and you can make a better interface for yourselves :star:

**Pull request merged**